bug: false positive on G118

We are running into a false positive with G118. The context cancel func is called when the Env.Shutdown chain is called (e.g. when the app receives SIGTERM), but gosec doesn't seem to think so. 

Code: 
https://github.com/moov-io/infra/pull/321/changes#diff-1802c2d91c6911f7a32160f87f8976e6a2365d4c93829ce954fa248445685392

```
../../../../../../../Users/adam/code/moov-io/infra/tests/code/withcontext.go:32:39: G118: context cancellation function returned by WithCancel/WithTimeout/WithDeadline is not called (gosec)
	ctx, cancelFunc := context.WithCancel(ctx)
	                                     ^
1 issues:
```

Github Action: https://github.com/moov-io/infra/actions/runs/22771875810/job/66054828112?pr=321

--- a/analyzers/context_propagation.go
+++ b/analyzers/context_propagation.go
@@ -725,6 +725,14 @@ func isCancelCalled(cancelValue ssa.Value, allFuncs []*ssa.Function) bool {
 				if r.X == current {
 					queue = append(queue, r)
 				}
+			case *ssa.Return:
+				// Cancel function is returned to the caller — responsibility
+				// is transferred; treat as "called".
+				for _, result := range r.Results {
+					if result == current {
+						return true
+					}
+				}
 			}
 		}
 	}

--- a/testutils/g118_samples.go
+++ b/testutils/g118_samples.go
@@ -324,7 +324,7 @@ func multiContext(ctx context.Context) {
 }
 `}, 1, gosec.NewConfig()},
 
-	// Vulnerable: cancel returned to caller (analyzer cannot verify caller will use it)
+	// Safe: cancel returned to caller — responsibility is transferred
 	{[]string{`
 package main
 
@@ -333,7 +333,7 @@ import "context"
 func createContext(ctx context.Context) (context.Context, context.CancelFunc) {
 	return context.WithCancel(ctx)
 }
-`}, 1, gosec.NewConfig()},
+`}, 0, gosec.NewConfig()},
 
 	// Note: simple goroutines with Background() not detected when request param unused
 	{[]string{`
@@ -1509,4 +1509,47 @@ func multipleViolations(ctx context.Context) {
 	_, _, _ = cancel1, cancel2, cancel3
 }
 `}, 3, gosec.NewConfig()},
+
+	// Safe: cancel returned as func() and called by caller (issue #1584)
+	{[]string{`
+package main
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+)
+
+type Env struct {
+	DB       *sql.DB
+	Shutdown func()
+}
+
+func withContext(ctx context.Context, env *Env) error {
+	db, closeFn, err := initDatabase(ctx)
+	if err != nil {
+		return fmt.Errorf("creating database: %w", err)
+	}
+
+	prev := env.Shutdown
+	env.Shutdown = func() {
+		prev()
+		closeFn()
+	}
+
+	env.DB = db
+	return nil
+}
+
+func initDatabase(ctx context.Context) (*sql.DB, func(), error) {
+	ctx, cancelFunc := context.WithCancel(ctx)
+	_ = ctx
+
+	db, err := sql.Open("sqlite", "testing")
+	if err != nil {
+		return nil, cancelFunc, fmt.Errorf("opening database: %%w", err)
+	}
+	return db, cancelFunc, nil
+}
+`}, 0, gosec.NewConfig()},
 }

